File: adapters/flutter/lib/src/identity.dart
final RegExp _safeComponent = RegExp(r'^[a-z0-9_.-]+$');
final RegExp _unsafeChars = RegExp(r'[^a-z0-9_.-]');

String normalizeComponent(String raw) {
  var value = raw.toLowerCase().replaceAll(_unsafeChars, '_');
  while (value.contains('__')) {
    value = value.replaceAll('__', '_');
  }
  value = value.replaceAll(RegExp(r'^[._]+'), '');
  if (value.isEmpty) {
    value = '_';
  }
  return value;
}

bool isSafeComponent(String value) =>
    _safeComponent.hasMatch(value) &&
    !value.startsWith('.') &&
    !value.contains('__');

String encodeVariant(Map<String, String> variant) {
  final keys = variant.keys.toList()..sort();
  return keys.map((k) => '$k=${variant[k]}').join(',');
}

Map<String, String> decodeVariant(String encoded) {
  if (encoded.isEmpty) {
    return const {};
  }
  final result = <String, String>{};
  for (final pair in encoded.split(',')) {
    final index = pair.indexOf('=');
    if (index <= 0) {
      throw FormatException('Invalid variant pair: $pair');
    }
    result[pair.substring(0, index)] = pair.substring(index + 1);
  }
  return result;
}

class MitameIdentity {
  MitameIdentity({
    required this.platform,
    required this.group,
    required this.name,
    Map<String, String> variant = const {},
  }) : variant = Map.unmodifiable(variant) {
    for (final component in [platform, ...group, name]) {
      if (!isSafeComponent(component)) {
        throw ArgumentError('Unsafe identity component: $component');
      }
    }
    for (final entry in variant.entries) {
      if (!isSafeComponent(entry.key) || !isSafeComponent(entry.value)) {
        throw ArgumentError('Unsafe variant: ${entry.key}=${entry.value}');
      }
    }
  }

  final String platform;
  final List<String> group;
  final String name;
  final Map<String, String> variant;

  String get stem =>
      variant.isEmpty ? name : '${name}__${encodeVariant(variant)}';

  String get id => [platform, ...group, stem].join('/');

  String get relativePath => '$id.png';

  String get sidecarRelativePath => '$id.json';
}

File: adapters/android/mitame/src/main/kotlin/io/github/mataku/mitame/Identity.kt
package io.github.mataku.mitame

internal object Identity {
    private val camelBoundary = Regex("([a-z0-9])([A-Z])")
    private val acronymBoundary = Regex("([A-Z])([A-Z][a-z])")
    private val unsafe = Regex("[^a-z0-9_.-]")

    fun normalize(raw: String): String {
        var value = raw.replace(camelBoundary, "$1_$2").replace(acronymBoundary, "$1_$2").lowercase().replace(unsafe, "_")
        while (value.contains("__")) {
            value = value.replace("__", "_")
        }
        value = value.trimStart('.', '_')
        return value.ifEmpty { "_" }
    }

    fun encodeVariant(variant: Map<String, String>): String =
        variant.entries.sortedBy { it.key }.joinToString(",") { "${it.key}=${it.value}" }

    fun stem(name: String, variant: Map<String, String>): String =
        if (variant.isEmpty()) name else "${name}__${encodeVariant(variant)}"

    fun id(platform: String, group: List<String>, stem: String): String =
        (listOf(platform) + group + stem).joinToString("/")
}
